File: Python/Hard/2197_Replace_Non_Coprime_Numbers_In_Array/solution.py
class Solution:
    def replaceNonCoprimes(self, nums: list[int]) -> list[int]:
        stack = []
        for i in range(len(nums)):
            current = nums[i]
            while len(stack) != 0 and not Solution.isCoprime(stack[-1], current):
                num = stack.pop()
                current = Solution.LCM(num, current)
            stack.append(current)
        
        return stack
 

    def isCoprime(a, b):
        return Solution.gcd(a, b) == 1

    def LCM(a, b):
        g = Solution.gcd(a, b)
        return a // g * b

    def gcd(a, b):
        while b != 0:
            a, b = b, a%b
        return a

File: Golang/Hard/2197_Replace_Non_Coprime_Numbers_in_Array/solution.go
package replacenoncoprimenumbersinarray

func replaceNonCoprimes(nums []int) []int {
	stack := []int{}
	for _, num := range nums {
		current := num
		for len(stack) != 0 && !isCoprime(stack[len(stack)-1], current) {
			index := len(stack) - 1
			lastNum := stack[index]
			stack = stack[:index]
			current = LCM(lastNum, current)
		}
		stack = append(stack, current)
	}
	return stack
}

func isCoprime(a int, b int) bool {
	return gcd(a, b) == 1
}

func LCM(a int, b int) int {
	g := gcd(a, b)
	return a / g * b
}

func gcd(a int, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}
